gluonnlp/model/sequence_sampler.py

```diff
@@ -278,6 +289,81 @@ def hybrid_forward(self, F, samples, valid_length, log_probs, scores, step, beam
                chosen_word_ids, beam_alive_mask, new_states
 
 
+class _SamplingStepUpdate(HybridBlock):
+    def __init__(self, beam_size, eos_id, temperature=1.0, prefix=None, params=None):
+        super(_SamplingStepUpdate, self).__init__(prefix, params)
+        self._beam_size = beam_size
+        self._eos_id = eos_id
+        self._temperature = temperature
+        assert eos_id >= 0, 'eos_id cannot be negative! Received eos_id={}'.format(eos_id)
+
+    # pylint: disable=arguments-differ
+    def hybrid_forward(self, F, samples, valid_length, outputs, scores, beam_alive_mask, states):
+        """
+        Parameters
+        ----------
+        F
+        samples : NDArray or Symbol
+            The current samples generated by beam search. Shape (batch_size, beam_size, L)
+        valid_length : NDArray or Symbol
+            The current valid lengths of the samples
+        outputs: NDArray or Symbol
+            Decoder output (unnormalized) scores of the current step.
+            Shape (batch_size * beam_size, V)
+        scores : NDArray or Symbol
+            The previous scores. Shape (batch_size, beam_size)
+        beam_alive_mask : NDArray or Symbol
+            Shape (batch_size, beam_size)
+        states : nested structure of NDArrays/Symbols
+            Inner NDArrays have shape (batch_size * beam_size, ...)
+
+        Returns
+        -------
+        new_samples : NDArray or Symbol
+            The updated samples. Shape (batch_size, beam_size, L + 1)
+        new_valid_length : NDArray or Symbol
+            Valid lengths of the samples. Shape (batch_size, beam_size)
+        new_scores : NDArray or Symbol
+            Shape (batch_size, beam_size)
+        chosen_word_ids : NDArray or Symbol
+            The chosen word ids of the step. Shape (batch_size, beam_size). If it's negative,
+            no word will be appended to the beam.
+        beam_alive_mask : NDArray or Symbol
+            Shape (batch_size, beam_size)
+        new_states : nested structure of NDArrays/Symbols
+            Inner NDArrays have shape (batch_size * beam_size, ...)
+        """
+        beam_size = self._beam_size
+        # outputs: (batch_size, beam_size, vocab_size)
+        outputs = outputs.reshape(shape=(-4, -1, beam_size, 0))
+        smoothed_probs = (outputs / self._temperature).softmax(axis=2)
+        log_probs = F.log_softmax(outputs, axis=2).reshape(-3, -1)
+
+        # (batch_size, beam_size)
+        chosen_word_ids = F.sample_multinomial(smoothed_probs).astype('float32')
+        chosen_word_log_probs = log_probs[mx.nd.arange(log_probs.shape[0]),
+                                          chosen_word_ids.reshape(-1)] \
+                                .reshape_like(chosen_word_ids)
```

Comment:
We'd better use int32 for indexing. Float32 may overflow if there are a large number of words.